Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp

```diff
@@ -401,30 +481,97 @@ void createSeedsForGroupSycl(
       // *********** TRIPLET SEARCH - BEGIN *********** //
       //************************************************//
 
-      const auto maxMemoryAllocation = std::min(
-          edgesComb, globalBufferSize / uint32_t((sizeof(detail::TripletData) +
-                                                  sizeof(detail::SeedData)) *
-                                                 2));
+      // EXPLANATION OF INDEXING (second part)
+      /*
+        For the triplet search, we calculate the upper limit of constructible
+        triplets.
+
+        For this, we multiply the number of compatible bottom and compatible
+        top SPs for each middle SP, and add these together.
+        (nb0*nt0 + nb1*nt1 + ... where nbk is the number of compatible bottom
+        SPs for the kth middle SP, similarly ntb is for tops)
+
+        sumBotTopCombined is a prefix sum array (of length M+1) of the
+        calculated combinations.
 
-      detail::TripletData* deviceCurvImpact =
-          cl::sycl::malloc_device<detail::TripletData>(maxMemoryAllocation,
-                                                       (*q));
+        sumBotTopCombined:
+        ________________________________________________________
+        |     |         |                   |     |  M         | M = number
+        |  0  | nb0*nt0 | nb0*nt0 + nb1*nt1 | ... |  ∑ nbi+nti | of middle
+        |_____|_________|___________________|_____|_i=0________| space points
+
+        We will start kernels and reserve memory for these combinations but
+        only so much we can fit into memory at once.
+
+        We limit our memory usage to globalBufferSize/2, this is currently
+        hard-coded, but it could be configured. Actually, it would be better
+        to use a separate object that manages memory allocations and
+        deallocations and we could ask it to lend us as much memory as it is
+        happy to give.
+
+        For later, let maxMemoryAllocation be maximum allocatable memory for
+        triplet search.
+
+        We start by adding up summing the combinations, until we arrive at a
+        k which for:
+
+        k+1
+         ∑ nbi+nti > maxMemoryAllocation
+        i=0
+        (or k == M).
+
+        So we know, that we need to start our first kernel for the first k
+        middle SPs.
+
+        Inside the triplet search kernel we start with a binary search, to
```

Comment:
I don't understand why a binary search is necessary. Could you not calculate which thread corresponds to which middle SP?

Reply:
It is not necessary, just a bit cheaper (surprisingly) than constructing arrays of middle SP indices.